tfhe/src/core_crypto/entities/polynomial.rs

```diff
@@ -22,6 +24,104 @@ impl<T, C: ContainerMut<Element = T>> AsMut<[T]> for Polynomial<C> {
     }
 }
 
+impl<C: Container> Index<usize> for Polynomial<C> {
+    type Output = C::Element;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<usize> for Polynomial<C> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<Range<usize>> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: Range<usize>) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<Range<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: Range<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<RangeFrom<usize>> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: RangeFrom<usize>) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeFrom<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeFrom<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<RangeFull> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: RangeFull) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeFull> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeFull) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<RangeInclusive<usize>> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: RangeInclusive<usize>) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeInclusive<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeInclusive<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<RangeTo<usize>> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: RangeTo<usize>) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeTo<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeTo<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
+impl<C: Container> Index<RangeToInclusive<usize>> for Polynomial<C> {
+    type Output = [C::Element];
+
+    fn index(&self, index: RangeToInclusive<usize>) -> &Self::Output {
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeToInclusive<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeToInclusive<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
```

Comment:
Maybe it is possible to shorten these with 

```rust
impl<C, I> Index<I> for Polynomial<C>
where C: Container + Index<I> {
    type Output = <C as Index<I>>::Output;

    fn index(&self, index: I) -> &Self::Output {
        self.as_ref().index(index)
    }
}
```

and Same for IndexMut

Reply:
I'll check, but I tried a generic type name without bound it did not seem to work ok, maybe I wrote it wrong

Reply:
actually, I was trying to go through the self.as_ref() that returns a slice and the indexing seemed to not work ok the first time around, now looks to be fine, must have been a bit tired Friday, updating

Reply:
Yeah if you use .as_ref()/.as_mut the bounds would have to be `for <'a> &'a [C::Element]: Index<I>`

Reply:
looks like it was a bit easier than that, but this time I managed to get it working 🤷‍♂️